src/librustdoc/config.rs

```diff
@@ -417,7 +417,12 @@ impl Options {
             }
             return Err(0);
         }
+        let (lint_opts, describe_lints, lint_cap) = get_cmd_lint_options(matches, error_format);
 
+        if matches.free.is_empty() && lint_opts.is_empty() {
+            crate::usage("rustdoc");
+            return Err(0);
```

Comment:
Why was this change necessary? I'm pretty sure it had this behavior even before.

Reply:
Hi @jyn514 !
Apologies for late response, my dev env was broken.
If I just move the `get_cmd_lint_options`, I still get:
```
└─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustdoc 
error: missing file operand
```
So, I checked for the `matches` attributes as the condition leading to this message was checking whether `matches.free` was empty. I found out that `matches.free` was indeed empty even after this move but what was also different with this situation compared to other places where `matches.free` is expected to error out was that `lint_opts` was also empty.

Before this change,
```
└─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustdoc -W help
error: missing file operand
```
and 
```
└─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustdoc -W
error: Argument to option 'W' missing
```

After adding these condition checks,
```
└─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustdoc -W help
rustdoc [options] <input>

Options:
    -h, --help          show this help message
    -V, --version       print rustdoc's version
    -v, --verbose       use verbose output
    -r, --input-format [rust]
                        the input type of the specified file
    -w, --output-format [html]
                        the output type to write
    -o, --output PATH   where to place the output
        --crate-name NAME
                        specify the name of this crate
        --crate-type [bin|lib|rlib|dylib|cdylib|staticlib|proc-macro]
                        Comma separated list of types of crates
                        for the compiler to emit
    -L, --library-path DIR
                        directory to add to crate search path
        --cfg           pass a --cfg to rustc
        --extern NAME[=PATH]
```
and
```
└─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustdoc -W
error: Argument to option 'W' missing
```
Is it still incorrect behavior? I'm not sure what changed from before if it was working with just the move.

Reply:
> Is it still incorrect behavior?

You can test by checking against rustc :) rustdoc and rustc should behave the same for invalid arguments.

(I haven't checked myself, I can do so if you have trouble.)

Reply:
hmm I am a little confused now. It seems as if rustdoc already (i.e. without this patch) works like rustc.
For example:
```
└─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustc -L help
error: no input filename given
```
It does not however behave this way particularly for 4 options (out of which one is `-W`) which are mentioned in the "additional help section"
```
Additional help:
    -C help             Print codegen options
    -W help             Print 'lint' options and default settings
    -Z help             Print unstable compiler options
    --help -v           Print the full set of options rustc accepts
```
So, if I were to pass `help` as mentioned in the cmd output above with either of these 4 options, I'll get some help corresponding to that option, e.g.
```
└─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustc -W help

Available lint options:
    -W <foo>           Warn about <foo>
    -A <foo>           Allow <foo>
    -D <foo>           Deny <foo>
    -F <foo>           Forbid <foo> (deny <foo> and all attempts to override)


Lint checks provided by rustc:

                                                       name  default  meaning
                                                       ----  -------  -------
                     absolute-paths-not-starting-with-crate  allow    fully qualified paths that start with a module name instead of `crate`, `self`, or an extern crate name
                                               box-pointers  allow    use of owned (Box type) heap memory
                                       deprecated-in-future  allow    detects use of items that will be deprecated in a future version
                                  elided-lifetimes-in-paths  allow    hidden lifetime parameters in types are deprecated
```
But, it seems that rustdoc does not promise any additional help or has a separate document for any of its options. Also, `-W` seems to mean different things for rustc and rustdoc.

The only difference that I am able to see is that rustdoc by default also errors out
```
─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustdoc 
error: missing file operand
```
while rustc prints help
```
└─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustc
Usage: rustc [OPTIONS] INPUT

Options:
    -h, --help          Display this message
        --cfg SPEC      Configure the compilation environment
    -L [KIND=]PATH      Add a directory to the library search path. The
                        optional KIND can be one of dependency, crate, native,
                        framework, or all (the default).
    -l [KIND[:MODIFIERS]=]NAME[:RENAME]
                        Link the generated crate(s) to the specified native
                        library NAME. The optional KIND can be one of
                        static, framework, or dylib (the default).
                        Optional comma separated MODIFIERS
                        (bundle|verbatim|whole-archive|as-needed)
                        may be specified each with a prefix of either '+' to
                        enable or '-' to disable.
        --crate-type [bin|lib|rlib|dylib|cdylib|staticlib|proc-macro]
                        Comma separated list of types of crates
                        for the compiler to emit
        --crate-name NAME
```

Did I do what you asked me to check, @jyn514 ? If yes, where does it make us stand on this current rustdoc issue (problem with `-W help`)?

Reply:
> Is it still incorrect behavior? I'm not sure what changed from before if it was working with just the move.

Yes, it's still correct - it should be showing the same lint help as rustc for `-W help`, not the generic usage.

> But, it seems that rustdoc does not promise any additional help or has a separate document for any of its options. Also, -W seems to mean different things for rustc and rustdoc.

That should not be the case. If rustdoc doesn't give help for `-W help`, that's also a bug that should be fixed.

> The only difference that I am able to see is that rustdoc by default also errors out while rustc prints help

We should fix that too.

> hmm I am a little confused now. It seems as if rustdoc already (i.e. without this patch) works like rustc (for the -W help like scenarios).

I'm not sure why you think `-L` and `-W` are similar. `rustc --help` says
```
    -L [KIND=]PATH      Add a directory to the library search path. The
                        optional KIND can be one of dependency, crate, native,
                        framework, or all (the default).
...

Additional help:
    -C help             Print codegen options
    -W help             Print 'lint' options and default settings
    -Z help             Print unstable compiler options
    --help -v           Print the full set of options rustc accepts
```
`-L help` is a valid option that means "add the directory named help/ to the search path".